### PR TITLE
[FOR DISCUSSION] 21368-Integrate-FileAttributesPlugin

### PR DESCRIPTION
**Pull request by akgrant43**
For discussion only.

This PR merges in all the changes required to the image to support the (not so) new FileAttributesPlugin.

Note that this PR can't be reviewed by pulling the changes in to an existing image since it will attempt to perform file access in the middle of the import.   The updated image needs to be bootstrapped.

### CODE REVIEW

File: src/FileSystem-Core/DiskDirectoryEntry.class.st [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+{ #category : #'testing - stat' }
+DiskDirectoryEntry >> isBlock [
+	"Answer true if the receiver is a block device"
+	
+	^self primitives modeIsBlock: self mode.
+]
+
+{ #category : #'testing - stat' }
+DiskDirectoryEntry >> isCharacter [
+	"Answer true if the receiver is a character device"
```
Comment (jecisc) [inline src/FileSystem-Core/DiskDirectoryEntry.class.st]: What is a character device?

Comment (akgrant43) [reply src/FileSystem-Core/DiskDirectoryEntry.class.st]: Hi Cyril,

> What is a character device?

The example I'm aware of is /dev/urandom.  But not really having a better understanding, here's a couple of pages I found:

"A character device typically transfers data to and from a user application — they behave like pipes or serial ports, instantly reading or writing the byte data in a character-by-character stream. They provide the framework for many typical drivers, such as those that are required for interfacing to serial communications, video capture, and audio devices. "

- http://derekmolloy.ie/writing-a-linux-kernel-module-part-2-a-character-device/

https://unix.stackexchange.com/questions/37829/how-do-character-device-or-character-special-files-work also has quite a bit of information, e.g. that character devices don't support the normal seek and tell operations.

**Comment by pavel-krivanek** [comment]
the PR produces two new Undeclared (#WindowsFileAttributesPluginPrims #FileAttributePrimitives) and introduces new bad dependencies

**Comment by akgrant43** [comment]
Hi @pavel-krivanek ,

Thanks very much for your reply.

The two undeclared variables were in a no-longer-used method that
somehow got left behind.  I've removed it but will hold off pushing the
commit until I've addressed the bad dependencies.

I've also resolved the conflicts in FileReferenceTest (also still to be
pushed up).

Would you mind clarifying what the bad dependencies are?


Thanks again,
Alistair

**Comment by akgrant43** [comment]
A few comments on the CI tests, which reported 29 failures:

- 5 of the failures are reported twice - all the "New failing" are also "Existing failing".  Is that normal?
- I couldn't reproduce the Unix 32 ZipArchiveTest failures.
- The 14 DiskFileAttributesTests and FileReferenceAttributeTests failures are expected.  The changes require a VM dated 17 Jun or later, and the new VM was promoted just after the build started.  They should be green in the next run.

**Comment by akgrant43** [comment]
Some real issues to be addressed in the latest build...

**Comment by estebanlm** [comment]
ok, please reopen if still relevant :)

**Comment by akgrant43** [comment]
Hi Esteban,

Shall do.

I've since learnt more than I ever wanted to know about file name encoding on OSX.  I've also, I think, battled my way through file name encoding on Windows - functions that only accept ascii, functions that pretend to accept UTF8 but don't (I hate Windows), functions that require the "\\?\" prefix for long file names, functions that accept the "\\?\" prefix but then fail depending on the file name pattern (did I mention that I hate Windows?), ...

Anyway, I think I've got the VM working properly now (still to be committed), and am tidying up the smalltalk code.  After that I should retest all three platforms.  And then I can reopen the PR!

Cheers,
Alistair

**Comment by bencoman** [comment]
Hi Alistair,

Thanks for your effort and persistence with this.  I appreciate not needing
to personally plumb such depths of file name encoding if you can
successfully encapsulate it.

cheers -ben

**Comment by akgrant43** [comment]
Hi Ben,

Thanks for your encouragement.  I have to admit that I'll be happy when I can spend more time programming in Smalltalk! :-)

Cheers,
Alistair

**Comment by akgrant43** [comment]
This PR has been superseeded by #2431

**Pull request closed without merging**